services/sharedchannel/sync_send.go

```diff
@@ -0,0 +1,260 @@
+		select {
+		case <-scs.changeSignal:
+			for {
+				task, ok := scs.removeOldestTask()
+				if !ok {
+					break
+				}
+				if err := scs.processTask(task); err != nil {
+					scs.server.GetLogger().Error("Failed to update shared channel", mlog.String("channelId", task.channelId), mlog.String("remoteId", task.remoteId), mlog.Err(err))
+					// put task back into map so it will update again
+					scs.addTask(task)
+				}
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+// removeOldestTask removes and returns the oldest task in the task map.
+// If no tasks are available then false is returned.
+func (scs *Service) removeOldestTask() (syncTask, bool) {
+	scs.mux.Lock()
+	defer scs.mux.Unlock()
+
+	var oldestTask syncTask
+	var oldestKey string
+
+	for k, v := range scs.tasks {
+		if v.AddedAt.Before(oldestTask.AddedAt) || oldestTask.AddedAt.IsZero() {
+			oldestKey = k
+			oldestTask = v
+		}
+	}
+
+	if oldestKey != "" {
+		delete(scs.tasks, oldestKey)
+		return oldestTask, true
+	}
+	return oldestTask, false
+}
+
+// processTask updates one or more remote clusters with any new channel content.
+func (scs *Service) processTask(task syncTask) error {
+	var err error
+	var remotes []*model.RemoteCluster
+
+	if task.remoteId == "" {
+		filter := model.RemoteClusterQueryFilter{
+			InChannel:     task.channelId,
+			OnlyConfirmed: true,
+		}
+		remotes, err = scs.server.GetStore().RemoteCluster().GetAll(filter)
+		if err != nil {
+			return err
+		}
+	} else {
+		rc, err := scs.server.GetStore().RemoteCluster().Get(task.remoteId)
+		if err != nil {
+			return err
+		}
+		if !rc.IsOnline() {
+			return fmt.Errorf("Failed updating shared channel '%s' for offline remote cluster '%s'", task.channelId, rc.DisplayName)
+		}
+		remotes = []*model.RemoteCluster{rc}
+	}
+
+	cache := make(msgCache)
+	errs := merror.New()
+	for _, rc := range remotes {
+		if err := scs.updateForRemote(task.channelId, rc, cache); err != nil {
+			errs.Append(err)
+			// retry...
+			if task.retryCount < MaxRetries {
+				retryTask := newSyncTask(task.channelId, rc.RemoteId)
+				retryTask.retryCount = task.retryCount + 1
+				scs.addTask(retryTask)
+			}
+		}
+	}
+	return errs.ErrorOrNil()
+}
+
+// updateForRemote updates a remote cluster with any new posts/reactions for a specific
+// channel. If many changes are found, only the oldest X changes are sent and the channel
+// is re-added to the task map. This ensures no channels are starved for updates even if some
+// channels are very active.
+func (scs *Service) updateForRemote(channelId string, rc *model.RemoteCluster, cache msgCache) error {
+	scr, err := scs.server.GetStore().SharedChannel().GetRemoteByIds(channelId, rc.RemoteId)
+	if err != nil {
+		return err
+	}
+
+	opts := model.GetPostsSinceOptions{
+		ChannelId: channelId,
+		Time:      scr.LastSyncAt,
+	}
+	posts, err := scs.server.GetStore().Post().GetPostsSince(opts, true)
+	if err != nil {
+		return err
+	}
+
+	var repeat bool
+
+	pSlice := posts.ToSlice()
```

Comment:
This might be a good time to micro-optimize `PostList.ToSlice()` by pre-allocating the array to a fixed length/capacity.

Reply:
Done.